File: packages/snow_draw_core/lib/draw/actions/history_coalescing.dart
import 'package:meta/meta.dart';

/// Coalescing hint for high-frequency actions that should collapse history.
///
/// Actions with the same [key] dispatched within [window] are merged into a
/// single undo/redo entry.
@immutable
class HistoryCoalescing {
  const HistoryCoalescing({
    required this.key,
    this.window = const Duration(milliseconds: 220),
  }) : assert(key != '', 'key must not be empty');

  /// Stable group key used to match adjacent actions for coalescing.
  final String key;

  /// Maximum time gap allowed between adjacent actions in this group.
  final Duration window;
}

/// Optional interface for actions that want history coalescing behavior.
mixin HistoryCoalescingProvider {
  /// Returns a coalescing hint, or `null` to disable coalescing.
  HistoryCoalescing? get historyCoalescing => null;
}

File: packages/snow_draw_rust_engine/src/draw/actions/history_coalescing.rs
#![allow(dead_code)]

use std::time::Duration;

/// Coalescing hint for high-frequency actions that should collapse history.
///
/// Adjacent actions with the same key dispatched inside `window` can be merged
/// into a single history record.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct HistoryCoalescing {
    /// Stable group key used to match adjacent actions.
    pub key: String,
    /// Maximum time gap allowed between adjacent actions in this group.
    pub window: Duration,
}

impl HistoryCoalescing {
    /// Default coalescing window (220ms), matching Dart defaults.
    pub const DEFAULT_WINDOW: Duration = Duration::from_millis(220);

    pub fn new(key: impl Into<String>) -> Self {
        Self::with_window(key, Self::DEFAULT_WINDOW)
    }

    pub fn with_window(key: impl Into<String>, window: Duration) -> Self {
        let key = key.into();
        assert!(!key.is_empty(), "key must not be empty");

        Self { key, window }
    }
}

/// Optional interface for actions that expose history coalescing hints.
pub trait HistoryCoalescingProvider {
    /// Returns coalescing metadata, or `None` to disable coalescing.
    fn history_coalescing(&self) -> Option<&HistoryCoalescing> {
        None
    }
}
